src/client/resources/web/js/message_types.mjs

```diff
@@ -0,0 +1,62 @@
+// @ts-check
+'use strict';
+
+/**
+ * Server information matching ChatServerInfo on server
+ * @typedef {Object} ServerInfo
+ * @property {string} name
+ * @property {string} identifier
+ */
+
+/**
+ * Base message interface matching WebsocketJsonMessage on server
+ * @typedef {Object} BaseModServerMessage
+ * @property {number} timestamp
+ * @property {ServerInfo} server
+ * @property {string} minecraftVersion
+ */
+
+/**
+ * Chat message from Minecraft
+ * @typedef {Object} ChatMessage
+ * @property {'chatMessage'} type
+ * @property {import('./message_parsing.mjs').Component} payload
+ */
+
+/**
+ * @typedef {BaseModServerMessage & ChatMessage} ModServerMessage
+ */
+
+/**
+ * Minimal type guard for TypeScript type narrowing.
+ * No complete validation because the server (mod) and client (this) are tightly coupled in one package.
+ * @param {unknown} message
+ * @returns {message is ModServerMessage}
+ */
+export function isModServerMessage(message) {
+    if (typeof message !== 'object' || message === null) {
+        return false;
+    }
+    if (!('type' in message)) {
+        return false;
+    }
+
+    /** @type {any} */
+    const messageWithType = message;
```

Comment:
I'd do this as:

```typescript
return typeof message.type === 'string' && message.type === 'chatMessage';
```

No need to break out the `any` (which is generally to be avoided, even though it's not going to cause trouble here).